parser-typechecker/src/Unison/Runtime/Machine.hs

```diff
@@ -279,150 +279,152 @@ exec !env !denv !ustk !bstk !k (BPrim1 LOAD i) = do
+  atomicEval env trackThreads (poke bstk) c
   pure (denv, ustk, bstk, k)
 {-# inline exec #-}
 
-eval :: CCache -> DEnv
+eval :: CCache -> DEnv -> (ThreadId -> IO ())
      -> Stack 'UN -> Stack 'BX -> K -> Section -> IO ()
-eval !env !denv !ustk !bstk !k (Match i (TestT df cs)) = do
+eval !env !denv !trackThreads !ustk !bstk !k (Match i (TestT df cs)) = do
   t <- peekOffBi bstk i
-  eval env denv ustk bstk k $ selectTextBranch t df cs
-eval !env !denv !ustk !bstk !k (Match i br) = do
+  eval env denv trackThreads ustk bstk k $ selectTextBranch t df cs
+eval !env !denv !trackThreads !ustk !bstk !k (Match i br) = do
   n <- peekOffN ustk i
-  eval env denv ustk bstk k $ selectBranch n br
-eval !env !denv !ustk !bstk !k (Yield args)
+  eval env denv trackThreads ustk bstk k $ selectBranch n br
+eval !env !denv !trackThreads !ustk !bstk !k (Yield args)
   | asize ustk + asize bstk > 0 , BArg1 i <- args = do
-    peekOff bstk i >>= apply env denv ustk bstk k False ZArgs
+    peekOff bstk i >>= apply env denv trackThreads ustk bstk k False ZArgs
   | otherwise = do
     (ustk, bstk) <- moveArgs ustk bstk args
     ustk <- frameArgs ustk
     bstk <- frameArgs bstk
-    yield env denv ustk bstk k
-eval !env !denv !ustk !bstk !k (App ck r args) =
+    yield env denv trackThreads ustk bstk k
+eval !env !denv !trackThreads !ustk !bstk !k (App ck r args) =
   resolve env denv bstk r
-    >>= apply env denv ustk bstk k ck args
-eval !env !denv !ustk !bstk !k (Call ck n args) =
+    >>= apply env denv trackThreads ustk bstk k ck args
+eval !env !denv !trackThreads !ustk !bstk !k (Call ck n args) =
   combSection env (CIx dummyRef n 0)
-    >>= enter env denv ustk bstk k ck args
-eval !env !denv !ustk !bstk !k (Jump i args) =
-  peekOff bstk i >>= jump env denv ustk bstk k args
-eval !env !denv !ustk !bstk !k (Let nw cix) = do
+    >>= enter env denv trackThreads ustk bstk k ck args
+eval !env !denv !trackThreads !ustk !bstk !k (Jump i args) =
+  peekOff bstk i >>= jump env denv trackThreads ustk bstk k args
+eval !env !denv !trackThreads !ustk !bstk !k (Let nw cix) = do
   (ustk, ufsz, uasz) <- saveFrame ustk
   (bstk, bfsz, basz) <- saveFrame bstk
-  eval env denv ustk bstk (Push ufsz bfsz uasz basz cix k) nw
-eval !env !denv !ustk !bstk !k (Ins i nx) = do
-  (denv, ustk, bstk, k) <- exec env denv ustk bstk k i
-  eval env denv ustk bstk k nx
-eval !_   !_    !_    !_    !_ Exit = pure ()
-eval !_   !_    !_    !_    !_ (Die s) = die s
+  eval env denv trackThreads ustk bstk (Push ufsz bfsz uasz basz cix k) nw
+eval !env !denv !trackThreads !ustk !bstk !k (Ins i nx) = do
+  (denv, ustk, bstk, k) <- exec env denv trackThreads ustk bstk k i
+  eval env denv trackThreads ustk bstk k nx
+eval !_   !_    !_    !_trackThreads !_    !_ Exit = pure ()
+eval !_   !_    !_    !_trackThreads !_    !_ (Die s) = die s
 {-# noinline eval #-}
 
-forkEval :: CCache -> Closure -> IO ThreadId
-forkEval env clo
-  = forkIO (apply1 err env clo)
```

Comment:
For anyone wondering, this is the source of the issue.
`forkIO` creates a new thread that's not linked to the current thread, so async exceptions aren't propagated and so if the threads don't exit on their own they'll exist as long as the parent process does, which sometimes is UCM, so they outlive their "run" so to speak.